microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java

```diff
@@ -0,0 +1,116 @@
+import static org.eclipse.lsp4mp.jdt.internal.faulttolerance.MicroProfileFaultToleranceConstants.FALLBACK_METHOD_FALLBACK_ANNOTATION_MEMBER;
+import static org.eclipse.lsp4mp.jdt.internal.faulttolerance.java.MicroProfileFaultToleranceErrorCode.FALLBACK_METHOD_DOES_NOT_EXIST;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.IAnnotation;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.IMemberValuePair;
+import org.eclipse.jdt.core.IMethod;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.ITypeRoot;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4mp.jdt.core.java.diagnostics.IJavaDiagnosticsParticipant;
+import org.eclipse.lsp4mp.jdt.core.java.diagnostics.JavaDiagnosticsContext;
+import org.eclipse.lsp4mp.jdt.core.utils.JDTTypeUtils;
+import org.eclipse.lsp4mp.jdt.core.utils.PositionUtils;
+
+public class MicroProfileFaultToleranceDiagnosticsParticipant implements IJavaDiagnosticsParticipant {
+
+	private static final Pattern fallbackMethodNameExtractor = Pattern.compile(FALLBACK_METHOD_FALLBACK_ANNOTATION_MEMBER + "\\s*=\\s*([^,]+),");
+	
+	@Override
+	public boolean isAdaptedForDiagnostics(JavaDiagnosticsContext context, IProgressMonitor monitor)
+			throws CoreException {
+		IJavaProject javaProject = context.getJavaProject();
+		return JDTTypeUtils.findType(javaProject, FALLBACK_ANNOTATION) != null;
+	}
+
+	@Override
+	public List<Diagnostic> collectDiagnostics(JavaDiagnosticsContext context, IProgressMonitor monitor)
+			throws CoreException {
+		ITypeRoot root = context.getTypeRoot();
+		IJavaElement[] elements = root.getChildren();
+		List<Diagnostic> diagnostics = new ArrayList<>();
+		collectDiagnostics(elements, diagnostics, context, monitor);
+		return diagnostics;
+	}
+
+	private void collectDiagnostics(IJavaElement[] elements, List<Diagnostic> diagnostics,
+			JavaDiagnosticsContext context, IProgressMonitor monitor) {
+		for (IJavaElement element : elements) {
+			if (monitor.isCanceled()) {
+				return;
+			} else if (element.getElementType() == IJavaElement.TYPE) {
+				IType type = (IType) element;
+				validateClass(type, diagnostics, context, monitor);
+			}
+		}
+	}
+
+	private void validateClass(IType type, List<Diagnostic> diagnostics, JavaDiagnosticsContext context,
+			IProgressMonitor monitor) {
+		try {
+			IMethod[] methods = type.getMethods();
+			Set<String> methodSet = Stream.of(methods).map(m -> {
+				return m.getElementName();
+			}).collect(Collectors.toUnmodifiableSet());
+			for (IMethod method : methods) {
```

Comment:
@datho7561 I have quickly play with JDT CompilationUnit DOM (so please clean my code if you wish to use it). Here my code:

```java
CompilationUnit ast = ASTResolving.createQuickFixAST(type.getCompilationUnit(), null);
			ast.accept(new ASTVisitor() {

				@Override
				public boolean visit(MethodDeclaration node) {
					List modifiers = node.modifiers();
					for (Object modifier : modifiers) {
						if (modifier instanceof NormalAnnotation) {
							List values = ((NormalAnnotation) modifier).values();
							for (Object value : values) {
								if (value instanceof MemberValuePair) {
									MemberValuePair pair = (MemberValuePair) value;
									String n = pair.getName().getFullyQualifiedName();
									if ("fallbackMethod".equals(n)) {
										Expression expr = pair.getValue();
										String v = expr.toString();
										String fallbackMethod = v.substring(1, v.length() - 1);
										if (!methodSet.contains(fallbackMethod)) {
											try {
												IOpenable openable = type.getCompilationUnit();

												diagnostics.add(context.createDiagnostic(context.getUri(),
														"The method '" + fallbackMethod + "' does not exist",
														context.getUtils().toRange(openable,
																pair.getValue().getStartPosition(),
																pair.getValue().getLength()),
														DIAGNOSTIC_SOURCE, FALLBACK_METHOD_DOES_NOT_EXIST));
											} catch (JavaModelException e) {
												// TODO Auto-generated catch block
												e.printStackTrace();
											}
										}

									}
								}
							}
						}
					}
					super.visit(node);
					return true;
				}
			});
```

Reply:
I managed to get it to work without it. It turns out I was using the wrong name for the Fallback annotation.

Reply:
I agree, the ASTVisitor is the way to go if the range should cover the value of `fallbackMethod`. Converting to draft while I work on this.

Reply:
I switched over to ASTVisitor, using Angelo's code, cleaning it up a bit.